### DISCUSSION TITLE
Default parameter values

### DISCUSSION THREAD

**Discussion by david-ragazzi** [modular/modular#1151] [Q&A] [answered]
In [changelog](https://docs.modular.com/mojo/changelog.html#v0.4.0-for-mac-2023-10-19), it's said that Mojo now supports default parameter values...

...in functions:

```
fn foo[a: Int = 3, msg: StringLiteral = "woof"]():
    print(msg, a)

fn main():
    foo()  # prints 'woof 3'
    foo[5]()  # prints 'woof 5'
    foo[7, "meow"]()  # prints 'meow 7'
```

and in structs:

```
@value
struct DefaultParams[msg: StringLiteral = "woof"]:
    alias message = msg

fn main():
    print(DefaultParams[]().message)  # prints 'woof'
    print(DefaultParams["meow"]().message)  # prints 'meow'

``` 

However this is not the way that you specificy default parameters in Python!

Look this example:

``` 
def foo(a, msg="woof"):  # 'a' is not optional, so it doesn't have a default value
    print(a, msg)

foo(1)
foo(2, "meow")
``` 

In the example above, you see that Python put optional parameters after non-optional parameter and all of them in the same parentesis. 

Look this: https://www.geeksforgeeks.org/default-arguments-in-python/

The same code of the changelog but in the correct Python syntax:

```
fn foo(a: Int = 3, msg: StringLiteral = "woof"):
    print(msg, a)

fn main():
    foo()  # prints 'woof 3'
    foo(5)  # prints 'woof 5'
    foo(7, "meow")  # prints 'meow 7'
```

In structs:

```
@value
struct DefaultParams(msg: StringLiteral = "woof"):
    alias message = msg

fn main():
    print(DefaultParams().message)  # prints 'woof'
    print(DefaultParams("meow").message)  # prints 'meow'
``` 

The way that it's now is fully imcompatilble with Python. Is there a (really strong) reason for that?

**Comment by Mogball** [comment, marked as answer]
I think you're mixing "parameters" and "arguments". The stuff in between square brackets are "parameter values", i.e. compile-time constants. You can still specify default arguments the same way as in Python. E.g. the code you posted

```python
fn foo(a: Int = 3, msg: StringLiteral = "woof"):
    print(msg, a)

fn main():
    foo()  # prints 'woof 3'
    foo(5)  # prints 'woof 5'
    foo(7, "meow")  # prints 'meow 7'
```

Is also valid Mojo code.

**Reply by david-ragazzi** [reply]
I got it. Thank you for clarify it.
